Pendapatan Royalti

michaelr1300 opened the pull request:
## Migrations
- Tabel Royalty (Catat riwayat royalti)
- Tabel Payment_Royalty (Catat data pembayaran royalti penulis: pembayaran royalti terakhir, status royalti saat ini)
- Update kolom royalti tabel invoice_book (Tambah default value untuk royalti = 15%)
- Tambah kolom royalti di tabel book (Default = 15%, untuk simpan data royalti buku)

## Pendapatan
- Fitur pendapatan hanya untuk menampilkan rekap pendapatan dari faktur, user cukup pilih fitur pendapatan, lalu klik bar pada grafik
- Pendapatan hanya mengambil data dari faktur yang sudah `finish` atau `cancel`
- Data harga faktur yang statusnya `cancel` akan dibuat menjadi 0 (permintaan dari staff ugm press)

## Royalti
- Halaman index_royalti hanya menampilkan daftar penulis yang bukunya terjual selama periode royalti
- Royalti diatur lewat halaman book_form (seperti mengatur harga buku dan berat buku)
- Batas periode royalti maks kemarin (Supaya faktur hari ini dan faktur2 masa depan tidak dianggap sudah terbayar royaltinya padahal faktur belum muncul)

### Alur pembayaran royalti:
1. Masuk ke halaman royalti lalu pilih penulis yang akan dibayar royaltinya
2. Pilih periode pembayaran royalti
3. Klik ajukan royalti (Data royalti akan dibuat sesuai periode dengan status `requested`)
4. Setelah dibayar klik tombol bayar 
5. Masukkan ID transaksi (Data royalti akan diubah statusnya menjadi `paid`)

## Video Pendapatan
Index pendapatan

https://user-images.githubusercontent.com/46708593/121072530-89599400-c7fb-11eb-88fb-a1102e637219.mp4

Detail pendapatan (Klik bar pada chart)

https://user-images.githubusercontent.com/46708593/121072551-91193880-c7fb-11eb-9cc4-b100b99efa0a.mp4

Faktur yang dicancel harganya akan dibuat jadi 0

https://user-images.githubusercontent.com/46708593/121072847-ece3c180-c7fb-11eb-896e-d442ac957fee.mp4

## Video Royalti

Alur pembuatan royalti

https://user-images.githubusercontent.com/46708593/121072952-1b619c80-c7fc-11eb-8d3b-3aca7aeb0d96.mp4

Riwayat Royalti

https://user-images.githubusercontent.com/46708593/121072970-261c3180-c7fc-11eb-93e4-0cc815950ca3.mp4

michaelr1300 committed:
Fitur Pendapatan

michaelr1300 committed:
Update Invoice_model.php

esterinac committed:
Update invoice model

fix book request filter

michaelr1300 committed:
Fitur Royalty

michaelr1300 commented:
Maaf mas kali ini agak banyak file changed nya. Dari fitur sendiri sebenarnya sederhana, hanya mengambil data dari faktur kemudian ditotal. Untuk yang royalti sama prinsipnya, hanya dikelompokkan per penulis.

bagaskarala commented:
maksudnya periode royalti tuh gimana ya?

bagaskarala commented:
beda royalti di tabel book dan invoice_book apa?

bagaskarala commented:
bisa dikasih link ke faktur
![image](https://user-images.githubusercontent.com/14112952/121276933-cc019600-c8f9-11eb-976d-d9c7b1f5013f.png)

bagaskarala commented:
semua yang rupiah perlu diformat Rp, dan pie chart ga perlu pakai garis-garis grid di belakangnya
![image](https://user-images.githubusercontent.com/14112952/121276996-e6d40a80-c8f9-11eb-9e63-62cdcd90b2f7.png)

michaelr1300 commented:
> maksudnya periode royalti tuh gimana ya?

Periode royalti itu batas pembayarannya mas. Jadi misal periode royalti 1 Mei 2021 - 31 Mei 2021 berarti penulis dapat royalti dari buku yang terjual di tanggal itu.

michaelr1300 commented:
> beda royalti di tabel book dan invoice_book apa?

Ini sama sih mas, cuma dicatet di invoice book untuk jaga-jaga jika suatu saat royaltinya berubah. Sama seperti harga, kalau hanya disimpan di tabel book nanti kalau harga berubah harga di faktur juga jadi berubah. Royalti juga begitu

bagaskarala commented:
rupiah penjualan ngambil dari data mana ya? harga buku tidak 0 tapi penjualannya kok 0
![image](https://user-images.githubusercontent.com/14112952/121278116-269bf180-c8fc-11eb-9ba2-113b166baadb.png)

bagaskarala commented:
daftar nama penerima royalti difilter sampai hari kemarin kan? tapi kalo diganti manual di url bakalan tampil, mohon dikasih validasi ya
![image](https://user-images.githubusercontent.com/14112952/121278983-ed648100-c8fd-11eb-99b5-473c2b0129ce.png)

bagaskarala commented:
tabel `royalty_payment` fungsinya untuk apa ya? ku liat isinya hampir sama dgn `royalty`, mungkin gak kalo dijadikan satu aja di tabel `royalty`?

michaelr1300 committed:
Fix Earning UI

- Tambah link ke faktur di detail pendapatan
- Hilangkan X Y axis di pie chart

michaelr1300 commented:
> rupiah penjualan ngambil dari data mana ya? harga buku tidak 0 tapi penjualannya kok 0
> ![image](https://user-images.githubusercontent.com/14112952/121278116-269bf180-c8fc-11eb-9ba2-113b166baadb.png)

Datanya dari kolom `price` di tabel `invoice_book` mas. Bukan dr harga di tabel book

michaelr1300 committed:
Update Royalty

- Compact database table
- Validation filter
- Bug fix

michaelr1300 committed:
Royalty Route

bagaskarala commented on application/modules/royalty/views/view_royalty.php:
```diff
@@ -0,0 +1,458 @@
+                                class="font-weight-bold"
+                            >
+                                Akhir Periode Royalti Sebelumnya</label>
+                            <!-- <?php if ($pending_royalty) : ?> -->
```
ini dipake atau engga kodenya, kalo enga, dihapus ya

michaelr1300 committed:
Add Spacing

bagaskarala commented on application/modules/earning/controllers/Earning.php:
```diff
@@ -0,0 +1,170 @@
+            'excel'         => $this->input->get('excel', true)
+        ];
+        if ($filters['date_year'] == NULL && $filters['invoice_type'] == NULL) {
+            $filters['date_year'] = '2021';
```
jangan hard code ya

bagaskarala commented on application/modules/earning/controllers/Earning.php:
```diff
@@ -0,0 +1,170 @@
+            'excel'         => $this->input->get('excel', true)
+        ];
+        if ($filters['date_year'] == NULL && $filters['invoice_type'] == NULL) {
+            $filters['date_year'] = '2021';
```
hardcode (2)

bagaskarala commented on application/modules/royalty/controllers/Royalty.php:
```diff
@@ -0,0 +1,192 @@
+            $latest_royalty->details = $this->royalty->author_details($author_id, $latest_filters)[0];
+        }
+
+        if ($latest_royalty == NULL) $last_paid_date = '2021/01/01';
```
ini hardcode juga

bagaskarala commented on application/modules/royalty/controllers/Royalty.php:
```diff
@@ -0,0 +1,192 @@
+        $latest_royalty = $this->royalty->fetch_latest_royalty($author_id);
+        //jika belum ada data royalti
+        if ($latest_royalty == NULL) {
+            $last_paid_date = '2021/01/01';
```
harcode

bagaskarala commented on application/modules/royalty/models/Royalty_model.php:
```diff
@@ -0,0 +1,109 @@
+    public function author_details($author_id, $filters)
+    {
+        $last_paid_date = '';
+        if ($filters['last_paid_date'] == '') $last_paid_date = "2021/01/01";
```
harcode

bagaskarala commented on application/modules/royalty/models/Royalty_model.php:
```diff
@@ -0,0 +1,109 @@
+    public function stocks_info($author_id, $filters)
+    {
+        $last_paid_date = '';
+        if ($filters['last_paid_date'] == '') $last_paid_date = "2021/01/01";
```
harcode

michaelr1300 commented:
> tabel `royalty_payment` fungsinya untuk apa ya? ku liat isinya hampir sama dgn `royalty`, mungkin gak kalo dijadikan satu aja di tabel `royalty`?

Untuk tabel `royalty_payment` sebenarnya untuk mencatat royalti terbaru dari penulis mas. Sekarang sudah diupdate jadi lebih compact hanya memakai tabel `royalty`. Sudah ditest dan fungsinya tetap sama saja mas, bisa di drop saja table royalty_payment, lalu di table royalty di empty.

Beberapa video untuk membantu testing:

- Case 1: Royalti diajukan sekaligus, lalu dibayar

https://user-images.githubusercontent.com/46708593/121366810-fa1bc000-c963-11eb-84e9-6b31732446b7.mp4

- Case 2: Royalti author dengan buku lebih dari 1

https://user-images.githubusercontent.com/46708593/121366955-16b7f800-c964-11eb-8ce5-fb85017a1cff.mp4

- Case 3: Royalti dibayar dalam beberapa periode

https://user-images.githubusercontent.com/46708593/121367027-26cfd780-c964-11eb-9c45-034f27e20d5a.mp4

bagaskarala commented:
banyak yg masih hardcode 1 januari 2021 di backend dan frontend, dibikin dinamis ya

michaelr1300 commented:
> banyak yg masih hardcode 1 januari 2021 di backend dan frontend, dibikin dinamis ya

Dibuat jadi 1 januari tahun ini gitu gpp mas? Jadi bsk kalau dah 2022 kalau null diubah jadi 1 januari 2022. Karena kemarin default kalau null kan dari SQL jadi 1970, dari Pak Wayan diminta jadi 2021 saja

bagaskarala commented:
> > tabel `royalty_payment` fungsinya untuk apa ya? ku liat isinya hampir sama dgn `royalty`, mungkin gak kalo dijadikan satu aja di tabel `royalty`?
> 
> Untuk tabel `royalty_payment` sebenarnya untuk mencatat royalti terbaru dari penulis mas. Sekarang sudah diupdate jadi lebih compact hanya memakai tabel `royalty`. Sudah ditest dan fungsinya tetap sama saja mas, bisa di drop saja table royalty_payment, lalu di table royalty di empty.
> 
> Beberapa video untuk membantu testing:
> 
> * Case 1: Royalti diajukan sekaligus, lalu dibayar
> 
>  bandicam.2021-06-09.20-41-49-702.mp4 
> * Case 2: Royalti author dengan buku lebih dari 1
> 
>  bandicam.2021-06-09.20-42-17-789.mp4 
> * Case 3: Royalti dibayar dalam beberapa periode
> 
>  bandicam.2021-06-09.20-42-49-918.mp4

video ke-3 detik ke-21, royalti belum dibayar tapi sudah keluar tanggal 1 januari ? dan detik ke-42 harusnya tanggal dibayar 9 juni kan bukan 1 januari?

bagaskarala commented:
> > banyak yg masih hardcode 1 januari 2021 di backend dan frontend, dibikin dinamis ya
> 
> Dibuat jadi 1 januari tahun ini gitu gpp mas? Jadi bsk kalau dah 2022 kalau null diubah jadi 1 januari 2022. Karena kemarin default kalau null kan dari SQL jadi 1970, dari Pak Wayan diminta jadi 2021 saja

atau dikosongi aja, nanti biar dipilih sendiri oleh user jika data awalnya masih kosong, tapi divalidasi juga agar data yang dikirim ada isian tanggalnya, begitu gimana?

michaelr1300 commented:
> > > tabel `royalty_payment` fungsinya untuk apa ya? ku liat isinya hampir sama dgn `royalty`, mungkin gak kalo dijadikan satu aja di tabel `royalty`?
> > 
> > 
> > Untuk tabel `royalty_payment` sebenarnya untuk mencatat royalti terbaru dari penulis mas. Sekarang sudah diupdate jadi lebih compact hanya memakai tabel `royalty`. Sudah ditest dan fungsinya tetap sama saja mas, bisa di drop saja table royalty_payment, lalu di table royalty di empty.
> > Beberapa video untuk membantu testing:
> > 
> > * Case 1: Royalti diajukan sekaligus, lalu dibayar
> > 
> > bandicam.2021-06-09.20-41-49-702.mp4
> > 
> > * Case 2: Royalti author dengan buku lebih dari 1
> > 
> > bandicam.2021-06-09.20-42-17-789.mp4
> > 
> > * Case 3: Royalti dibayar dalam beberapa periode
> > 
> > bandicam.2021-06-09.20-42-49-918.mp4
> 
> video ke-3 detik ke-21, royalti belum dibayar tapi sudah keluar tanggal 1 januari ? dan detik ke-42 harusnya tanggal dibayar 9 juni kan bukan 1 januari?

Ohh maaf mas mungkin itu agak membuat bingung tabelnya. Maksudnya itu tanggal akhir periode sebelumnya. Jadi yang 1 januari itu karena sebelumnya blm ada royalti maka dianggap periode terakhir nya jatuh tanggal 1 januari 2021.

bagaskarala commented:
> Ohh maaf mas mungkin itu agak membuat bingung tabelnya. Maksudnya itu tanggal akhir periode sebelumnya. Jadi yang 1 januari itu karena sebelumnya blm ada royalti maka dianggap periode terakhir nya jatuh tanggal 1 januari 2021.

menurutku mending kosong aja sih kalo memang sblmnya belom ada datanya

michaelr1300 committed:
Royalty Validation

michaelr1300 committed:
Fix Bug Validation

michaelr1300 commented:
Sudah ditambah validasi mas @bagaskarala mungkin bisa di cek

michaelr1300 committed:
Update UI index

bagaskarala commented:
@michaelr1300 untuk formatting rupiah di konsistenkan yah, misal Rp 2.500.000
![image](https://user-images.githubusercontent.com/14112952/121630667-c6e24980-caa7-11eb-84ef-f6fb38e24e87.png)

bagaskarala commented:
tanggal awal jadi hilang, jika tanggal akhir diganti. 
saranku mungkin  bisa diset aja via url `/royalty/view/yyy?start=aaa&end=bbb` biar fleksibel.

https://user-images.githubusercontent.com/14112952/121630923-4839dc00-caa8-11eb-8c74-66f28a77fe82.mp4

bagaskarala commented:
kalo gini gimana ya, tidak bisa bayar royalti

https://user-images.githubusercontent.com/14112952/121631998-5a1c7e80-caaa-11eb-8309-04b7c896b445.mp4

michaelr1300 committed:
Fix Royalty Earning

- Remove Hardcoded
- Fix royalty validation
- Fix earning formatting
- Update Royalty history (Seperti index, tapi menampilkan semua royalty. Index hanya menampilkan royalty yg belum dibayar)
- Route untuk pagination
- Sidebar untuk royalty history

michaelr1300 commented:
Hardcoded, Format rupiah, bug validation royalty sudah di fix mas @bagaskarala

bagaskarala commented:
ohiya btw aku malah jadi bingung ini mksudnya beda informasi yg ingin ditampilkan apa?
![image](https://user-images.githubusercontent.com/14112952/122077797-3f3d6780-ce26-11eb-9e0e-2d8ee2a1c93c.png)

menurutku kalo sumber datanya sama, misal ingin menampilkan perbedaan yang diajukan dan yang telah dibayar, mungkin pakai filter status aja atau pakai tab. kalau ada dua page gini isinya mirip malah membingungkan..

bagaskarala commented:
> Hardcoded, Format rupiah, bug validation royalty sudah di fix mas @bagaskarala

beberapa hardcode juga masih ada kalo ku search di https://github.com/bagaskarala/sigap/pull/51/files dengan keyword `2021`

michaelr1300 committed:
Rename Model Controller Proforma

michaelr1300 committed:
Update Royalty

- Ubah History jadi tab
- Update Sidebar: remove royalty child
- Filter history author dengan select2
- Redirect link author ke halaman author
- Tab Riwayat royalti di halaman author

michaelr1300 committed:
Remove Hardcoded

michaelr1300 commented:
Sudah diupdate mas @bagaskarala 
- Remove hardcoded `2021`
- Update view Royalty -> History dipindah jadi tab sendiri di sebelah index
- Update sidebar -> Remove child di menu royalty
- Update History -> Sudah di tab sendiri, filter author dengan dropdown
- Update Redirect -> Nama author jika di klik akan redirect ke halaman profil author

bagaskarala commented:
Tabel "tagihan royalti" harusnya disamakan dengan tabel di "riwayat royalti"
![image](https://user-images.githubusercontent.com/14112952/122368684-c0604000-cf87-11eb-975e-b7fbde1777b3.png)

Filternya juga seharusnya bentuknya sama
![image](https://user-images.githubusercontent.com/14112952/122368562-a6266200-cf87-11eb-89ab-2a0100244f55.png)

Behavior tombolnya juga harusnya sama, jika klik detail maka ke detail royalti, jika klik nama orang maka ke author detail
![image](https://user-images.githubusercontent.com/14112952/122368893-ee458480-cf87-11eb-90b7-65a45a718f3d.png)

Ini harusnya labelnya filter penulis
![image](https://user-images.githubusercontent.com/14112952/122369770-b1c65880-cf88-11eb-8eff-432dbda40f91.png)

bagaskarala commented:
Oiya untuk yg halaman tagihan royalti itu orangnya pasti cuma tampil sekali ya? Berarti gajadi pake dropdown, sudah bener pakai search box saja 👍 sorry aku salah paham

michaelr1300 committed:
Update UI Royalty

- Fix table inconsistency
- Fix filter inconsistency
- Fix redirect author
- Rename Button

michaelr1300 commented:
> Tabel "tagihan royalti" harusnya disamakan dengan tabel di "riwayat royalti"
> ![image](https://user-images.githubusercontent.com/14112952/122368684-c0604000-cf87-11eb-975e-b7fbde1777b3.png)
> 
> Filternya juga seharusnya bentuknya sama
> ![image](https://user-images.githubusercontent.com/14112952/122368562-a6266200-cf87-11eb-89ab-2a0100244f55.png)
> 
> Behavior tombolnya juga harusnya sama, jika klik detail maka ke detail royalti, jika klik nama orang maka ke author detail
> ![image](https://user-images.githubusercontent.com/14112952/122368893-ee458480-cf87-11eb-90b7-65a45a718f3d.png)
> 
> Ini harusnya labelnya filter penulis
> ![image](https://user-images.githubusercontent.com/14112952/122369770-b1c65880-cf88-11eb-8eff-432dbda40f91.png)

Oke mas sudah di perbaiki @bagaskarala

michaelr1300 commented:
Jadi gimana mas @bagaskarala ? Masih ada yg perlu diubah lagi atau sudah bisa di merge ya? Terima kasih

bagaskarala commented:
> Jadi gimana mas @bagaskarala ? Masih ada yg perlu diubah lagi atau sudah bisa di merge ya? Terima kasih

maaf baru sempet cek lagi, sip sementara sudah oke, nanti sambil ku tes2 lagi
